abel/basex.py

```diff
@@ -344,3 +536,144 @@ def generate_basis_sets(n, nbf='auto', verbose=False):
+
+    # the number of elements used to calculate the projected coefficeints
+    delta = np.fmax(np.arange(nbf_horz)*32 - MAX_BASIS_SET_OFFSET, MAX_BASIS_SET_OFFSET) 
+    for k in range(1, nbf_horz):
+        k2 = k*k # so we don't recalculate it all the time
+        log_k2 = log(k2) 
+        angn = exp(
+                    k2 * (1 - log_k2) + 
+                    gammaln(k2 + 0.5) - gammaln_0o5  
+                    # old form --> k2 - 2 * k2*log(k) +
+                    #              np.log(np.arange(0.5, k2 + 0.5)).sum()
+                    )
+        M_horz[Rm_h-1, k] =  2 * angn
+
+        for l in range(1, n_horz - Rm_h + 1):
+            l2 = l*l
+            log_l2 = log(l2)
+
+            val = exp(k2 * (1 + log(l2/k2)) - l2) 
+            # old form --> val = exp(k2 - l2 + 2 * k2*log((1.0 * l)/k)) 
+
+            Mc_horz[Rm_h - 1 + l, k] = val # All rows below center
+            Mc_horz[Rm_h - 1 - l, k] = val # All rows above center
+
+            aux = val + angn * Mc_horz[Rm_h - 1 + l, 0]
+
+            p = np.arange(max(1, l2 - delta[k]), min(k2 - 1,  l2 + delta[k]) + 1)
+
+            # We use here the fact that for p, k real and positive
+            #
+            #  np.log(np.arange(p, k)).sum() == gammaln(k) - gammaln(p) 
+            #
+            # where gammaln is scipy.misc.gammaln (i.e. the log of the Gamma function)
+            #
+            # The following line corresponds to the vectorized third
+            # loop of the original BASIS2.m matlab file.
+
+            aux += np.exp(k2 - l2 - k2*log_k2 + p*log_l2
+                      + gammaln(k2 + 1) - gammaln(p + 1) 
+                      + gammaln(k2 - p + 0.5) - gammaln_0o5
+                      - gammaln(k2 - p + 1)
+                      ).sum()
+
+            # End of vectorized third loop
+
+            aux *= 2
+
+            M_horz[Rm_h - 1 + l, k] = aux # All rows below center
+            M_horz[Rm_h - 1 - l, k] = aux # All rows above center
+
+        if verbose and k % 50 == 0:
+            sys.stdout.write('...{}'.format(k))
+            sys.stdout.flush()
+
+    if verbose:
+        print("...{}".format(k+1))
+
+    ####################################    
+    # Axial functions
+    ####################################
+
+    Z2_h = np.arange(0, n_vert)**2
+
+    M_vert = np.zeros((n_vert, nbf_vert))
+    Mc_vert = np.zeros((n_vert, nbf_vert))
+
+    # M_vert[:,0] = 2*np.exp(-Z2_h)
+    Mc_vert[:,0] = np.exp(-Z2_h)
+
+    if verbose:
+        print('Generating vertical BASEX basis sets for n_vert = {}, nbf_vert = {}:\n'.format(n_vert, nbf_vert))
+        sys.stdout.write('0')
+        sys.stdout.flush()
+
+    # delta_v = np.fmax(np.arange(nbf_vert)*32 - 0, 8000) 
+
+    for k in range(1, nbf_vert):
+        k2 = k*k
+        log_k2 = log(k2)
+
+        for l in range(1, n_vert):
+            l2 = l*l
+            log_l2 = log(l2)
+
+            val = exp(k2 * (1 + log(l2/k2)) - l2) 
+
+            Mc_vert[l, k] = val 
```

Comment:
I think we can vectorize the inner loop here, which should be a bit faster (not sure if this part of the code matters for the performance),

    l =  np.arange(1, n_vert)
    l2 = l*l
    log_l2 = log(l2)
    Mc_vert[:,k] = exp(k2 * (1 + log(l2/k2)) - l2)